src/main/java/core/stack/MateStack.java

```diff
@@ -1,19 +1,70 @@
 package core.stack;
 
+import java.util.EmptyStackException;
+
 public class MateStack<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final int MAX_ARRAY_SIZE = Integer.MAX_VALUE;
+    private static final int DEFAULT_MULTIPLIER = 2;
+
+    private T[] elementData;
+    private int size;
+    private int currentMaxSize;
+
+    public MateStack() {
+        elementData = (T[]) new Object[DEFAULT_CAPACITY];
+        currentMaxSize = DEFAULT_CAPACITY;
+    }
+
     public void push(T value) {
+        if (value == null) {
+            return;
+        }
+        if (size == currentMaxSize) {
+            resize();
+        }
+
+        elementData[size] = value;
+        size++;
+    }
+
+    private void resize() {
+        if (size == elementData.length) {
+            int newCapacity = currentMaxSize * DEFAULT_MULTIPLIER;
+            if (newCapacity - MAX_ARRAY_SIZE > 0) {
+                newCapacity = MAX_ARRAY_SIZE;
+            }
 
+            T[] newArray = (T[]) new Object[newCapacity];
+            System.arraycopy(elementData, 0, newArray, 0, elementData.length);
```

Comment:
Array copying and reference swapping should be atomic, to avoid data corruption in case of concurrent access. Consider using 'Arrays.copyOf' method, which handles this automatically.